agent/agent.go

```diff
@@ -483,7 +486,8 @@ func (a *Agent) withSession(ctx context.Context, fn func(session *session) error
 //
 // If an error is returned, the operation should be retried.
 func (a *Agent) UpdateTaskStatus(ctx context.Context, taskID string, status *api.TaskStatus) error {
```

Comment:
Can we change this method to coordinate the batching rather than propagating the interface update through?

Reply:
The reason not to do so is because we end up having to locally cache batch members in more and more places, which is the pattern that has bit us and caused this bug in the first place. 

We'd end up with 1 task statuses map in the StatusReporter, and then another task status map to keep track of the batch in the Agent, and we have to synchronize access to both in their respective places, and it's just asking for trouble.

The interface change seems more impactful than it actually is because once I changed the interface, there was no reason not to go through and add the batch method up and down the stack. Like, once I added `UpdateTaskStatuses` on agent, i also added` UpdateTaskStatuses` on the status reporter because I already had it and there was the opportunity for a good optimization there.

Reply:
The maps have different functionality. One map is a per task register of the latest task. The other is a queue of statuses that need to be batched out. The rest of the code changes less if we just make this handle batching.